### PR TITLE
Relation selection and filtering

### LINKED ISSUE #448: Problem when querying ManyToMany relationships in hydra-3.1.0-alpha.7
Hydra version: `@joystream/hydra 3.1.0-alpha.7`
Warthog version: `@joystream/warthog 2.35.0`

**Schema:**
```graphql
type StorageBucket @entity {
  "Runtime bucket id"
  id: ID!

  # ...

  "Bags assigned to be stored by the bucket"
  storedBags: [StorageBag!] @derivedFrom(field: "storedBy")
}

type StorageBag @entity {
  "Storage bag id"
  id: ID!

  # ...

  "Storage buckets assigned to store the bag"
  storedBy: [StorageBucket!]
}
```

**Query:**
```graphql
storageBuckets {
    # ...
    storedBags {
      id
    }
}
```

**Error:**
```
      "message": "column storagebucket.storedbags does not exist",
```

**Additional notes:**

Works in earlier versions of hydra (for example `hydra 3.1.0-alpha.1`)

### PR DESCRIPTION
**Pull request by ondratra**
affects: @joystream/hydra-cli, @joystream/hydra-indexer-gateway

This PR adds a new feature - filtering by properties from related entities (solves #444). Fixes #448. Theoretically, it could also positively affect performance for some queries due to their simplification (assuming the code is correct and they generate the same results).

A significant amount of work related to this PR was done to Warthog's `BaseService` in Joystream/warthog#3 . Changes in Hydra mainly consist of bumping `Warthog` dependency version and simplifying some `WarthogBaseService` queries that can now leverage new Warthog's features.

### CODE REVIEW

**Comment by Lezek123** [comment]
> This PR adds a new feature - filtering by properties from related entities

Just to clarify - the filtering was already working in most cases, there was just some problem with `id_eq` queries.

## Test results

### Olympia mappings

Tests were performed on branch https://github.com/Joystream/joystream/pull/2547) after updating the dependencies


- **Issue https://github.com/Joystream/hydra/issues/444:**

    ```
    workers(where: { group: { id_eq: "storageWorkingGroup" } }) {
      id
    }
    ```
    **FIXED!**

    ```
    workers(where: { membership: { metadata: { name_eq: "Alice" } } }) {
      # ...
    }
    ```
    **NOT FIXED** (but afaik it's a non-blocking issue)

    ```
    query {
      workersConnection(where: { group: { id_eq: "forumWorkingGroup" } }) {
        totalCount
      }
    }
    ```
    **FIXED!**

There was another issues discovered during the tests, probably related to https://github.com/Joystream/hydra/pull/449:
```
ERROR: Stopping the proccessor due to errors: {} name: TypeError, message: Cannot read property 'forum_thread_id' of undefined, stack: TypeError: Cannot read property 'forum_thread_id' of undefined
    at sanitizeNullCharacter (/joystream/node_modules/@joystream/hydra-processor/lib/db/subscribers.js:37:16)
    at sanitizeEntity (/joystream/node_modules/@joystream/hydra-processor/lib/db/subscribers.js:30:9)
    at SanitizationSubscriber.beforeInsert (/joystream/node_modules/@joystream/hydra-processor/lib/db/subscribers.js:11:9)
```

After commenting out `sanitizeNullCharacter` call in `node_modules/@joystream/hydra-processor/lib/db/subscribers.js` all tests passed succesfully.

### Storage v2 mappings (https://github.com/Joystream/joystream/pull/2547):

Testes were performed on branch https://github.com/Joystream/joystream/pull/2582 after updating the dependencies.

- https://github.com/Joystream/hydra/issues/448 - appears to be **FIXED**
- There is still an ongoing isssue with many-to-many relationships: https://github.com/Joystream/hydra/issues/454

File: packages/hydra-cli/src/templates/graphql-server/src/WarthogBaseService.ts.mst [hunk 1]
```diff
@@ -16,146 +16,30 @@ export class WarthogBaseService<E extends BaseModel> extends BaseService<E> {
...
-
-      if (rest) {
-        processWheres(qb, rest);
-      }
-      return qb;
-    };
 
-    if (Object.keys(where).length) {
-      processWhereInput(qb, where);
+        qb = qb.addSelect(`${this.klass}.${field}`, aliases(field));
```
Comment (dzhelezov) [inline packages/hydra-cli/src/templates/graphql-server/src/WarthogBaseService.ts.mst:37]: Not sure if it's applicable, but please have a look at this commit: https://github.com/subsquid/hydra/commit/ca4bcd914926beceee8be79d376ce628f298ee7c

**Comment by ondratra** [comment]
Recent PR changes include dependency updates and change from dependency on NPM packages from `@dzlzv` namespace to `@joystream` namespace. 

Update in dependencies caused one weird error in variant filtering test.
```
The processor processed block 342
    1) should find by a variant filter

Disconnecting from the chain

  0 passing (42s)
  1 failing

  1) end-to-end transfer tests
       should find by a variant filter:
     Error: QueryBuilder parameter keys may only contain numbers, letters, underscores, or periods.-"account"."status"isTypeOf_eq-: {"response":{"errors":[{"message":"QueryBuilder parameter keys may only contain numbers, letters, underscores, or periods.-\"account\".\"status\"isTypeOf_eq-","locations":[{"line":3,"column":5}],"path":["accounts"],"extensions":{"code":"INTERNAL_SERVER_ERROR"}}],"data":null,"status":200,"headers":{}},"request":{"query":"\n  query {\n    accounts(where: { status_json: { isTypeOf_eq: \"Miserable\" } }) {\n      status {\n        __typename\n        ... on Miserable {\n          hates\n        }\n      }\n    }\n  }\n"}}
      at /mnt/work/Joystream/hydra/node_modules/graphql-request/src/index.ts:325:11
      at step (/mnt/work/Joystream/hydra/node_modules/graphql-request/dist/index.js:63:23)
      at Object.next (/mnt/work/Joystream/hydra/node_modules/graphql-request/dist/index.js:44:53)
      at fulfilled (/mnt/work/Joystream/hydra/node_modules/graphql-request/dist/index.js:35:58)
      at processTicksAndRejections (internal/process/task_queues.js:95:5)

```
The error was resolved by relying on the older `typeorm` version `0.2.34` and forcing it via `resolutions` directive in `codegen` generated `packaged.json`. Strangely enough, it seems that this test is working on Subsquid's fork of Hydra that is using `typeorm v0.2.34`, so we might remove the change to `codegen` when the forks are united.

File: packages/hydra-cli/src/templates/graphql-server/src/WarthogBaseService.ts.mst (hunk 1 above)
Comment (ondratra) [reply packages/hydra-cli/src/templates/graphql-server/src/WarthogBaseService.ts.mst:37]: This change in the referenced commit should be the same as in this PR, just formulated a bit differently.

**Pull request merged**